resources/js/beatmap-discussions/discussions-state.ts

```diff
@@ -1,14 +1,431 @@
+  }
+
+  constructor(private store: BeatmapsetDiscussionsStore, state?: string) {
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+    const existingState = state == null ? null : parseState(state);
+
+    if (existingState != null) {
+      Object.apply(this, existingState);
+    } else {
+      this.jumpToDiscussion = true;
+      for (const discussion of store.beatmapset.discussions) {
+        if (discussion.posts != null) {
+          for (const post of discussion.posts) {
+            this.readPostIds.add(post.id);
+          }
+        }
+      }
+    }
+
+    this.currentBeatmapId = (findDefault({ group: this.groupedBeatmaps }) ?? this.firstBeatmap).id;
+
+    // Current url takes priority over saved state.
+    const query = parseUrl(null, store.beatmapset.discussions);
+    if (query != null) {
+      // TODO: maybe die instead?
+      this.currentMode = query.mode;
+      this.currentFilter = query.filter;
+      if (query.beatmapId != null) {
+        this.currentBeatmapId = query.beatmapId;
+      }
+      this.selectedUserId = query.user ?? null;
+    }
 
-  constructor() {
     makeObservable(this);
   }
+
+  @action
+  changeDiscussionPage(page?: string) {
+    if (!isDiscussionPage(page)) return;
+
+    const url = makeUrl({
+      beatmap: this.currentBeatmap,
+      filter: this.currentFilter,
+      mode: page,
+      user: this.selectedUserId ?? undefined,
+    });
+
+    if (page === 'events') {
+      // record page and filter when switching to events
+      this.previousPage = this.currentMode;
+      this.previousFilter = this.currentFilter;
+    } else if (this.currentFilter !== this.previousFilter) {
+      // restore previous filter when switching away from events
+      this.currentFilter = this.previousFilter;
+    }
+
+    this.currentMode = page;
+    Turbolinks.controller.advanceHistory(url);
+  }
+
+  @action
+  changeFilter(filter: unknown) {
+    if (!isFilter(filter)) return;
+
+    // restore previous page when selecting a filter.
+    if (this.currentMode === 'events') {
+      this.currentMode = this.previousPage;
+    }
+
+    this.currentFilter = filter;
+  }
+
+  @action
+  changeGameMode(mode: GameMode) {
+    const beatmap = findDefault({ items: this.groupedBeatmaps.get(mode) });
+    if (beatmap != null) {
+      this.currentBeatmapId = beatmap.id;
+    }
+  }
+
+  discussionsByBeatmap(beatmapId: number) {
+    return this.presentDiscussions.filter((discussion) => (discussion.beatmap_id == null || discussion.beatmap_id === beatmapId));
+  }
+
+  @action
+  markAsRead(ids: number | number[]) {
+    if (Array.isArray(ids)) {
+      ids.forEach((id) => this.readPostIds.add(id));
+    } else {
+      this.readPostIds.add(ids);
+    }
+  }
+
+  toJsonString() {
+    return JSON.stringify(toJS(this), (key, value) => {
+      // don't serialize constructor dependencies, they'll be handled separately.
+      if (key === 'beatmapset' || key === 'store') {
```

Comment:
It's going to have problems with object depth and circular references as well - `stringify` here should only be for properties the class itself owns and can avoid issues with.

Reply:
not sure what you're saying there 🤔 

also

> `stringify` here should only be for properties the class itself owns and can avoid issues with.

currently any new properties are automatically stringified and there's no indication that the existing properties must be in specific type to avoid the issues.